File: dimos/mapping/hyperspace/embedder.py
# Copyright 2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""SigLIP2 with a text-aligned embedding per patch, not just per image.

SigLIP's raw patch tokens are not aligned with its text tower; only the
attention-pooled image vector is. Running every patch token through that
pooling head as its own length-1 sequence (the MaskCLIP trick) gives 576
vectors per 384x384 image that *are* text-aligned, and those are what
hyperspace stores and scores.

Two families of checkpoint are handled. Fixed-resolution ones
(``siglip2-<size>-patch<p>-<px>``) squash the image to a square and give a
``side x side`` grid. NaFlex ones (``siglip2-<size>-patch16-naflex``) keep the
aspect ratio and give an ``h x w`` grid whose size depends on the image and on a
patch budget; ``<id>@<budget>`` names one with its budget, e.g.
``google/siglip2-base-patch16-naflex@576``.

:class:`PatchEnsemble` runs several checkpoints over the same frame. Each
one hallucinates in its own places (the floor next to a cone for one, a
reflection for another), so a per-cell *minimum* over their scores keeps
what they agree on and drops the rest (sweep of 2026-09-11, plan.md step 7).
"""

from __future__ import annotations

from functools import cached_property
from typing import TYPE_CHECKING, Any, Literal

import numpy as np
import torch
from torch.nn import functional
from transformers import SiglipModel, SiglipTextModel, SiglipVisionModel

from dimos.models.embedding.siglip import SigLIPModel, SigLIPModelConfig

if TYPE_CHECKING:
    from numpy.typing import NDArray

    from dimos.msgs.sensor_msgs.Image import Image

# Fixed-resolution SigLIP2: 384 / 16 = 24 patches per side = 576 patches, 1152-d.
SIGLIP2_MODEL_NAME = "google/siglip2-so400m-patch16-384"
# The ensemble the size sweep picked: two small checkpoints whose per-cell
# minimum beats the single So400m above on precision and recall at 2-5x the
# frame rate and 60% of the memory (sf_office, 302 keyframes, 4 queries).
DEFAULT_MEMBERS = ["google/siglip2-base-patch16-224", "google/siglip2-base-patch16-256"]


def parse_member(spec: str) -> tuple[str, int | None]:
    """``"google/siglip2-base-patch16-naflex@576"`` -> (id, 576); no ``@`` -> (id, None)."""
    name, _, budget = spec.partition("@")
    return name.strip(), int(budget) if budget else None


def member_tag(spec: str) -> str:
    """A short, file-safe name: ``google/siglip2-base-patch16-naflex@576`` ->
    ``base-patch16-naflex-576``. Stored with every keyframe so the query side
    knows which text towers to load."""
    name, budget = parse_member(spec)
    tag = name.rstrip("/").rsplit("/", 1)[-1].removeprefix("siglip2-")
    return f"{tag}-{budget}" if budget else tag


class SigLIP2PatchesConfig(SigLIPModelConfig):
    model_name: str = SIGLIP2_MODEL_NAME
    # Which tower(s) to load. The full model is 4.3 GB in f32; a module that
    # only embeds images (the patch writer) or only text (the query side)
    # loads about half of that with "vision" or "text".
    towers: Literal["both", "vision", "text"] = "both"
    # Kept for callers that predate `towers`; same as towers="text".
    text_only: bool = False
    # NaFlex checkpoints only: patches per image (the aspect ratio is kept, so
    # an 848x480 frame gets 18x32 = 576 of them). None = the checkpoint's own
    # default, 256. Ignored by fixed-resolution checkpoints.
    max_patches: int | None = None


class SigLIP2Patches(SigLIPModel):
    """SigLIP2 that also returns one text-aligned embedding per image patch."""

    config: SigLIP2PatchesConfig

    @property
    def towers(self) -> str:
        return "text" if self.config.text_only else self.config.towers

    @cached_property
    def naflex(self) -> bool:
        """NaFlex checkpoints are a different architecture (``model_type``
        "siglip2"): flattened patches through a linear embed, a position grid
        resized per image, and a patch budget instead of a fixed square."""
        from transformers import AutoConfig

        return AutoConfig.from_pretrained(self.config.model_name).model_type == "siglip2"

    @cached_property
    def _model(self) -> Any:  # type: ignore[override]
        self._ensure_cuda_initialized()
        if self.naflex:
            from transformers import Siglip2Model, Siglip2TextModel, Siglip2VisionModel

            loaders = {"both": Siglip2Model, "vision": Siglip2VisionModel, "text": Siglip2TextModel}
        else:
            loaders = {"both": SiglipModel, "vision": SiglipVisionModel, "text": SiglipTextModel}
        return (
            loaders[self.towers]
            .from_pretrained(self.config.model_name)
            .eval()
            .to(self.config.device)
        )

    @cached_property
    def _processor(self) -> Any:  # type: ignore[override]
        # AutoProcessor picks the Siglip2 (NaFlex) processor when the
        # checkpoint needs it; the base class hard-codes the SigLIP one.
        from transformers import AutoProcessor

        return AutoProcessor.from_pretrained(self.config.model_name)

    @cached_property
    def _vision(self) -> Any:
        """The vision transformer (with its attention-pooling head), whichever class holds it."""
        if self.towers == "text":
            raise RuntimeError("SigLIP2Patches loaded only the text tower; it cannot embed images")
        return self._model.vision_model

    @cached_property
    def _vision_config(self) -> Any:
        config = self._model.config
        return getattr(config, "vision_config", config)

    @cached_property
    def patches_per_side(self) -> int:
        """Side of the square patch grid. Fixed-resolution checkpoints only;
        a NaFlex grid depends on the image, see :meth:`embed_grids`."""
        if self.naflex:
            raise RuntimeError(f"{self.config.model_name} is NaFlex: its grid depends on the image")
        vision = self._vision_config
        return int(vision.image_size // vision.patch_size)

    @cached_property
    def dim(self) -> int:
        return int(self._vision_config.hidden_size)

    def embed_patches(self, *images: Image) -> list[NDArray[np.float32]]:
        """Per-patch L2-normalized embeddings, one ``[patches, dim]`` array per image."""
        return [grid for grid, _ in self.embed_grids(*images)]

    def embed_grids(self, *images: Image) -> list[tuple[NDArray[np.float32], tuple[int, int]]]:
        """Per-patch embeddings with the grid they form: ``([rows*cols, dim],
        (rows, cols))`` per image, row-major."""
        from PIL import Image as PILImage

        pil_images = [PILImage.fromarray(img.to_rgb().data) for img in images]
        if not self.naflex:
            with torch.inference_mode():
                inputs = self._processor(images=pil_images, return_tensors="pt")
                grids = self.patches_from_pixels(inputs["pixel_values"])
            side = round(grids[0].shape[0] ** 0.5) if grids else 0
            return [(grid, (side, side)) for grid in grids]
        # NaFlex frames of one aspect ratio share a grid, but the processor
        # pads every image to the budget; one image per call keeps the
        # bookkeeping simple and the cost is the same.
        out = []
        for image in pil_images:
            extra = (
                {}
                if self.config.max_patches is None
                else {"max_num_patches": self.config.max_patches}
            )
            with torch.inference_mode():
                inputs = self._processor(images=[image], return_tensors="pt", **extra).to(
                    self.config.device
                )
                vision = self._vision
                hidden = vision(
                    pixel_values=inputs["pixel_values"].to(next(vision.parameters()).dtype),
                    attention_mask=inputs["pixel_attention_mask"],
                    spatial_shapes=inputs["spatial_shapes"],
                ).last_hidden_state
                rows, cols = (int(v) for v in inputs["spatial_shapes"][0])
                hidden = hidden[0, : rows * cols]
                per_patch = vision.head(hidden.reshape(rows * cols, 1, hidden.shape[-1]))
                per_patch = functional.normalize(per_patch.reshape(rows * cols, -1).float(), dim=-1)
            out.append(
                (np.ascontiguousarray(per_patch.cpu().numpy(), dtype=np.float32), (rows, cols))
            )
        return out

    def patches_from_pixels(self, pixel_values: torch.Tensor) -> list[NDArray[np.float32]]:
        """Same as :meth:`embed_patches` from an already preprocessed ``[B,3,S,S]``
        tensor (fixed-resolution checkpoints)."""
        vision = self._vision
        with torch.inference_mode():
            pixel_values = pixel_values.to(
                self.config.device, dtype=next(vision.parameters()).dtype
            )
            hidden = vision(pixel_values=pixel_values).last_hidden_state
            batch, patches, dim = hidden.shape
            # MaskCLIP: each patch token pooled alone, so it lands in the text-aligned space.
            per_patch = vision.head(hidden.reshape(batch * patches, 1, dim)).reshape(
                batch, patches, dim
            )
            per_patch = functional.normalize(per_patch.float(), dim=-1).cpu().numpy()
        return [np.ascontiguousarray(grid, dtype=np.float32) for grid in per_patch]

    # SigLIP2 was trained on 64-token, max-length padded text; its tokenizer
    # snapshot does not always carry that length, so pass it.
    TEXT_LENGTH = 64

    def embed_text_array(self, *texts: str) -> NDArray[np.float32]:
        """L2-normalized text embeddings as a ``[len(texts), dim]`` array."""
        if self.towers == "vision":
            raise RuntimeError("SigLIP2Patches loaded only the vision tower; it cannot embed text")
        with torch.inference_mode():
            inputs = self._processor(
                text=list(texts),
                return_tensors="pt",
                padding="max_length",
                max_length=self.TEXT_LENGTH,
                truncation=True,
            ).to(self.config.device)
            if self.towers == "text":
                features = self._model(**inputs).pooler_output
            else:
                features = self._model.get_text_features(**inputs)
            features = functional.normalize(features.float(), dim=-1)
        return np.ascontiguousarray(features.cpu().numpy(), dtype=np.float32)


class PatchEnsemble:
    """Several :class:`SigLIP2Patches` over the same frames, one grid each.

    ``specs`` are checkpoint ids, NaFlex ones optionally with ``@<budget>``.
    The first member is the *primary*: its vectors also go into the store's
    vector index, so tools that only understand one grid per keyframe keep
    working. Scores are pooled per cell at query time (see
    ``HyperspaceQuery.pooled_hot_patches``); nothing here mixes embeddings.
    """

    def __init__(
        self, specs: list[str], device: str, towers: Literal["both", "vision", "text"] = "both"
    ) -> None:
        if not specs:
            raise ValueError("PatchEnsemble needs at least one checkpoint")
        self.specs = list(specs)
        self.tags = [member_tag(spec) for spec in specs]
        self.members = [
            SigLIP2Patches(model_name=name, max_patches=budget, device=device, towers=towers)
            for name, budget in map(parse_member, specs)
        ]

    @property
    def primary(self) -> SigLIP2Patches:
        return self.members[0]

    def start(self) -> None:
        for member in self.members:
            member.start()

    def stop(self) -> None:
        for member in self.members:
            member.stop()

    def embed_grids(self, image: Image) -> list[tuple[NDArray[np.float32], tuple[int, int]]]:
        """One ``(grid, (rows, cols))`` per member, in member order."""
        return [member.embed_grids(image)[0] for member in self.members]

    def embed_text_array(self, *texts: str) -> list[NDArray[np.float32]]:
        """One ``[len(texts), dim]`` array per member: each has its own text tower."""
        return [member.embed_text_array(*texts) for member in self.members]

    def embed_text(self, text: str) -> list[NDArray[np.float32]]:
        """The query-side callable: one vector per member."""
        return [vectors[0] for vectors in self.embed_text_array(text)]

File: dimos/mapping/hyperspace/rust/hyperspace/src/embedder.rs
// Copyright 2026 Dimensional Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Model backends behind traits so the core builds and tests without a GPU.

use crate::patch::{normalize, PatchGrid};
use crate::ImageFrame;
use std::collections::HashMap;

pub trait Embedder: Send {
    /// Per-patch, text-aligned, L2-normalized embeddings for one RGB frame.
    fn embed(&mut self, frame: &ImageFrame) -> Result<PatchGrid, String>;
    fn grid_shape(&self) -> (usize, usize);
    fn dim(&self) -> usize;
}

pub trait TextEmbedder: Send {
    /// L2-normalized text embedding in the same space as the patch embeddings.
    fn embed_text(&mut self, text: &str) -> Result<Vec<f32>, String>;
}

/// Deterministic pseudo-embeddings from image content (no model). Two frames
/// with identical pixels get identical grids; different pixels, different grids.
/// Good enough to exercise the pipeline, meaningless for real queries.
pub struct HashEmbedder {
    pub rows: usize,
    pub cols: usize,
    pub dim: usize,
}

impl Default for HashEmbedder {
    fn default() -> Self {
        HashEmbedder {
            rows: 24,
            cols: 24,
            dim: 32,
        }
    }
}

impl Embedder for HashEmbedder {
    fn embed(&mut self, frame: &ImageFrame) -> Result<PatchGrid, String> {
        let mut data = Vec::with_capacity(self.rows * self.cols * self.dim);
        let patch_h = (frame.height as usize / self.rows).max(1);
        let patch_w = (frame.width as usize / self.cols).max(1);
        for row in 0..self.rows {
            for col in 0..self.cols {
                let mut seed: u64 = 0xcbf29ce484222325;
                for y in
                    (row * patch_h..((row + 1) * patch_h).min(frame.height as usize)).step_by(4)
                {
                    for x in
                        (col * patch_w..((col + 1) * patch_w).min(frame.width as usize)).step_by(4)
                    {
                        let offset = (y * frame.width as usize + x) * 3;
                        for byte in &frame.data[offset..offset + 3] {
                            seed = (seed ^ (*byte as u64)).wrapping_mul(0x100000001b3);
                        }
                    }
                }
                let mut vector: Vec<f32> = (0..self.dim)
                    .map(|k| {
                        seed = seed
                            .wrapping_mul(6364136223846793005)
                            .wrapping_add(1442695040888963407 + k as u64);
                        ((seed >> 33) as f32 / (1u64 << 31) as f32) - 0.5
                    })
                    .collect();
                normalize(&mut vector);
                data.extend(vector);
            }
        }
        Ok(PatchGrid::from_f32(self.rows, self.cols, self.dim, &data))
    }

    fn grid_shape(&self) -> (usize, usize) {
        (self.rows, self.cols)
    }

    fn dim(&self) -> usize {
        self.dim
    }
}

/// Text embeddings looked up from a fixed table (tests, offline tools).
#[derive(Default)]
pub struct TableTextEmbedder {
    pub table: HashMap<String, Vec<f32>>,
}

impl TextEmbedder for TableTextEmbedder {
    fn embed_text(&mut self, text: &str) -> Result<Vec<f32>, String> {
        let mut vector = self
            .table
            .get(text)
            .cloned()
            .ok_or_else(|| format!("no table embedding for {text:?}"))?;
        normalize(&mut vector);
        Ok(vector)
    }
}

/// Scripted per-frame grids keyed by (camera_frame, timestamp) for tests; falls
/// back to a hash grid for unscripted frames.
#[derive(Default)]
pub struct ScriptedEmbedder {
    pub grids: HashMap<(String, u64), PatchGrid>,
    pub fallback: Option<HashEmbedder>,
    pub rows: usize,
    pub cols: usize,
    pub dim: usize,
}

impl ScriptedEmbedder {
    pub fn key(camera_frame: &str, timestamp: f64) -> (String, u64) {
        (camera_frame.to_string(), (timestamp * 1e6).round() as u64)
    }

    pub fn script(&mut self, camera_frame: &str, timestamp: f64, grid: PatchGrid) {
        self.grids.insert(Self::key(camera_frame, timestamp), grid);
    }
}

impl Embedder for ScriptedEmbedder {
    fn embed(&mut self, frame: &ImageFrame) -> Result<PatchGrid, String> {
        if let Some(grid) = self
            .grids
            .get(&Self::key(&frame.camera_frame, frame.timestamp))
        {
            return Ok(grid.clone());
        }
        match &mut self.fallback {
            Some(fallback) => fallback.embed(frame),
            None => Err(format!(
                "no scripted grid for {} @ {}",
                frame.camera_frame, frame.timestamp
            )),
        }
    }

    fn grid_shape(&self) -> (usize, usize) {
        (self.rows, self.cols)
    }

    fn dim(&self) -> usize {
        self.dim
    }
}
